### PR TITLE
BLD: use compiler flags in a more portable way

### LINKED ISSUE #16935: BUG: using msvc + meson to build scipy --> cl cannot be used to compile programs
### Describe your issue.

I'm investigating building on Windows on a local machine. Rather than using mingw-w64 from rtools I thought I'd start to try using a conda based environment, to see it's possible to setup an environment that would successfully build scipy.

I setup the environment and try to compile, but I get an issue with the (MSVC) C compiler straight away. It seems that a flag is being added, `/Wno-unused-function` that cl doesn't know about. Is this something that scipy could fix, or is this an upstream meson thing? I've tried building a simpler package and there were no issues:

```
:: this simple test works
mkdir meson-test
cd meson-test
meson init
meson build
```

(I'm also looking into using the `clang-cl` compiler supplied by the `compilers` conda-forge package, but can't get as far as building the simple test above).

### Reproducing Code Example

```python
conda create -n test
conda activate test
conda install -c conda-forge compilers python=3.10 meson

meson build
```

### Error message

```shell
Build started at 2022-08-31T10:31:58.903013
Main binary: C:\Users\anz\Miniconda3\envs\test\python.exe
Build Options: 
Python system: Windows
The Meson build system
Version: 0.63.1
Source dir: \\filer\home\anz\programming\scipy
Build dir: \\filer\home\anz\programming\scipy\build
Build type: native build
Project name: SciPy
Project version: 1.10.0.dev0
-----
Detecting compiler via: icl ""
-----
Detecting compiler via: cl /?
compiler returned <Popen: returncode: 0 args: ['cl', '/?']>
compiler stdout:
usage: cl [ option... ] filename... [ /link linkoption... ]

compiler stderr:
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27048 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.


Sanity testing C compiler: cl
Is cross compiler: False.
Sanity check compiler command line: cl sanitycheckc.c /Fesanitycheckc.exe /MD -Wno-unused-function -Wno-conversion -Wno-misleading-indentation -Wno-incompatible-pointer-types /nologo /showIncludes /utf-8 /link
Sanity check compile stdout:

-----
Sanity check compile stderr:
cl : Command line error D8021 : invalid numeric argument '/Wno-unused-function'

-----

\\filer\home\anz\programming\scipy\meson.build:1:0: ERROR: Compiler cl can not compile programs.
```

### SciPy/NumPy/Python version information

Python 3.10

### PR DESCRIPTION
**Pull request by rgommers**
Closes gh-16935

### CODE REVIEW

File: scipy/meson.build [hunk 1]
```diff
@@ -223,7 +222,16 @@ cython_gen_cpp = generator(cython_cli,
   output : '@BASENAME@.cpp',
   depends : [_cython_tree])
 
-c_undefined_ok = ['-Wno-maybe-uninitialized']
+# Check if compiler flags are supported. The standard convention used here is
+# to drop the leading dash, and turn remaining dashes into underscores
+Wno_maybe_uninitialized = []
+if cc.has_argument('-Wno-maybe-uninitialized')
+  Wno_maybe_uninitialized += ['-Wno-maybe-uninitialized']
+endif
```
Comment (rgommers) [inline scipy/meson.build:230]: @eli-schwartz does this look like the right thing to do (also style-wise), or is there a more idiomatic way?

**Comment by andyfaff** [comment]
@eli-schwartz mentioned `cxx.get_supported_arguments()`. If this was in Python you'd probably get a set of all the supported arguments, and then do an intersection with all the ones you wanted to add.

File: scipy/meson.build (hunk 1 above)
Comment (eli-schwartz) [reply scipy/meson.build:230]: Perhaps instead use:

```meson
Wno_maybe_uninitialized = cc.get_supported_arguments('-Wno-maybe-uninitialized')
Wno_discarded_qualifiers = cc.get_supported_arguments('-Wno-discarded-qualifiers')
```

:)

**Comment by rgommers** [comment]
I think I got them all now, this is ready. @andyfaff would be great if you can check that this solves the issue with MSVC.

**Comment by andyfaff** [comment]
It gets further into the sanity check, but then complains that the toolchain is too old:
```
(test) H:\programming\scipy>meson build
The Meson build system
Version: 0.63.1
Source dir: \\filer\home\anz\programming\scipy
Build dir: \\filer\home\anz\programming\scipy\build
Build type: native build
Project name: SciPy
Project version: 1.10.0.dev0
Activating VS 16.11.15
C compiler for the host machine: cl (msvc 19.29.30145 "Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30145 for x64")
C linker for the host machine: link link 14.29.30145.0
C++ compiler for the host machine: cl (msvc 19.29.30145 "Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30145 for x64")
C++ linker for the host machine: link link 14.29.30145.0
Host machine cpu family: x86_64
Host machine cpu: x86_64

\\filer\home\anz\programming\scipy\meson.build:30:4: ERROR: Problem encountered: SciPy requires at least vc142 (default with Visual Studio 2019) when building with MSVC
```

I've got VS2019 installed, so I would've thought it would use vc142 by default.
```
> mamba list
bzip2                     1.0.8                h8ffe710_4    conda-forge
ca-certificates           2022.6.15            h5b45459_0    conda-forge
libffi                    3.4.2                h8ffe710_5    conda-forge
libsqlite                 3.39.2               h8ffe710_1    conda-forge
libzlib                   1.2.12               h8ffe710_2    conda-forge
meson                     0.63.1                   pypi_0    pypi
openssl                   3.0.5                h8ffe710_1    conda-forge
pip                       22.2.2             pyhd8ed1ab_0    conda-forge
python                    3.9.13          hcf16a7b_0_cpython    conda-forge
setuptools                65.3.0             pyhd8ed1ab_1    conda-forge
sqlite                    3.39.2               h8ffe710_1    conda-forge
tk                        8.6.12               h8ffe710_0    conda-forge
tzdata                    2022c                h191b570_0    conda-forge
ucrt                      10.0.20348.0         h57928b3_0    conda-forge
vc                        14.2                 hb210afc_7    conda-forge
vs2015_runtime            14.29.30139          h890b9b1_7    conda-forge
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
xz                        5.2.6                h8d14728_0    conda-forge

```

**Review by ilayn — APPROVED** [review]

**Comment by rgommers** [comment]
@andyfaff okay, that's one step further. Maybe the version numbers aren't working the way I expected. It's this line in the top-level `meson.build` file:
```meson
if not cc.version().version_compare('>=1920')
```

You can put `message(cc.version())` right above it to print the actual version number.

Thanks for the testing, and thanks @ilayn for the review. This PR is complete, so let's merge it and let's leave further MSVC build issues to gh-16935.

**Comment by eli-schwartz** [comment]
> Maybe the version numbers aren't working the way I expected.

Isn't the version reported, `19.29.30145` rather than a number in the thousands?

**Pull request merged**